src/libraries/System.Security.Cryptography.Cose/src/System/Security/Cryptography/Cose/CoseSign1Message.cs

```diff
@@ -209,39 +209,57 @@ private static ReadOnlyMemory<byte> GetCoseAlgorithmFromProtectedHeaders(CoseHea
             // This authentication can be done either by placing the header in the protected header bucket or as part of the externally supplied data.
             if (!protectedHeaders.TryGetEncodedValue(CoseHeaderLabel.Algorithm, out ReadOnlyMemory<byte> encodedAlg))
             {
-                throw new CryptographicException(SR.Sign1SignAlgIsRequired);
+                throw new CryptographicException(SR.Sign1VerifyAlgIsRequired);
             }
 
             return encodedAlg;
         }
 
         // If we Validate: The caller did specify a COSE Algorithm, we will make sure it matches the specified key and hash algorithm.
-        // If we Add: The caller did not specify a COSE Algorithm, we will add the header for them, rather than throw. protectedHeaders will be altered in this case.
-        private static void ValidateOrAddAlgorithmHeader(CoseHeaderMap protectedHeaders, CoseHeaderMap unprotectedHeaders, KeyType keyType, HashAlgorithmName hashAlgorithm)
+        // If we Slip: The caller did not specify a COSE Algorithm, we will write the header for them, rather than throw.
+        private static void ValidateOrSlipAlgorithmHeader(
+            CoseHeaderMap protectedHeaders,
+            CoseHeaderMap unprotectedHeaders,
+            KeyType keyType,
+            HashAlgorithmName hashAlgorithm,
+            out int? headerValueToSlip)
         {
-            int algHeaderValue;
+            int algHeaderValue = GetCoseAlgorithmHeaderFromKeyTypeAndHashAlgorithm(keyType, hashAlgorithm);
 
             if (protectedHeaders.TryGetEncodedValue(CoseHeaderLabel.Algorithm, out ReadOnlyMemory<byte> encodedAlg))
             {
-                int? algHeaderValueNullable = DecodeCoseAlgorithmHeader(encodedAlg);
-                Debug.Assert(algHeaderValueNullable.HasValue, "Algorithm (alg) is a known header and should have been validated in Set[Encoded]Value()");
-                algHeaderValue = algHeaderValueNullable.Value;
-
-                HashAlgorithmName expectedHashAlgorithm = GetHashAlgorithmFromCoseAlgorithmAndKeyType(algHeaderValue, keyType);
-
-                if (expectedHashAlgorithm != hashAlgorithm)
-                {
-                    throw new CryptographicException(SR.Format(SR.Sign1SignHashAlgorithmDoesNotMatchSpecifiedAlg, hashAlgorithm.Name, algHeaderValue));
-                }
+                ValidateAlgorithmHeader(encodedAlg, algHeaderValue, keyType, hashAlgorithm);
+                headerValueToSlip = null;
+                return;
             }
 
             if (unprotectedHeaders.TryGetEncodedValue(CoseHeaderLabel.Algorithm, out _))
```

Comment:
Has the "no headers duplicated across protected & unprotected" already happened by this point?  If not, I'd do this check before the one for the protected headers.

Reply:
It already happened.